components/tikv_util/src/lib.rs

```diff
@@ -744,4 +769,22 @@ mod tests {
         });
         res.unwrap_err();
     }
+
+    #[test]
+    fn test_reserve_space_for_recover() {
+        let data_dir = Path::new("./");
+        file::delete_file_if_exist(data_dir.join(SPACE_PLACEHOLDER_FILE)).unwrap();
+        file::sync_dir(data_dir).unwrap();
+        let disk_stats_before = fs2::statvfs(data_dir).unwrap();
+        let cap1 = disk_stats_before.available_space();
+        let reserve_size = 64 * 1024;
+        reserve_space_for_recover(data_dir, reserve_size).unwrap();
+        let disk_stats_after = fs2::statvfs(data_dir).unwrap();
+        let cap2 = disk_stats_after.available_space();
+        assert_eq!(cap1 - cap2, reserve_size);
+        reserve_space_for_recover(data_dir, 0).unwrap();
+        let disk_stats = fs2::statvfs(data_dir).unwrap();
+        let cap3 = disk_stats.available_space();
+        assert_eq!(cap1, cap3);
```

Comment:
This test appears to be racy and should fail unexpectedly: it changes the size of individual files on disk then expects the total disk free space to be a specific number, but the test case does not control the free space for the entire disk.

I don't think the test can be written this way. Offhand I don't think the test can reliably use the available space in the filesystem without explicitly mounting its own filesystem that other processes won't modify. Without your own file system I think the best you can do is to measure the size of the written file.

Additionally, this test (and tests generally) should use a tempdir instead of writing to "./".

Reply:
Thanks for your suggestion. I will create a temp dir for test. If this file was created in a temp directory, there won't be another process change its available space, will it?
And I will add more comment for why I need to reserve  space for recover.

Reply:
Emmmm. It seems too hard to test this function without `available_space`. Do you have any better idea?

Reply:
> If this file was created in a temp directory, there won't be another process change its available space, will it?

By my understanding just using a temp directory is not enough to prevent the race condition. "available_space" reports space for the entire file system, and other processes may be modifying the rest of the file system at the same time this test is running.

Offhand, the best way I can think to test this reliably would be to make it a linux-only test, then create a temporary file system that can't be interfered with by other processes. e.g. mount a 1GB file as an ext3-formatted block device (somehow), and run the test inside _that_ filesystem.

Whether that is worth doing though I am not sure.

Reply:
It maybe too expensive for the unit test....